src/neoxp/Node/NodeUtility.cs

```diff
@@ -208,6 +212,186 @@ public static void SignOracleResponseTransaction(ProtocolSettings settings, Expr
+            // contractHash with the prefix 8. The following code uses stateAPI.GetStateAsync to retrieve
+            // the value with that key at the height state root and then deserializes it into a ContractState
+            // instance via GetInteroperable.
+
+            var key = new byte[21];
+            key[0] = 8; // ContractManagement.Prefix_Contract
+            _contractHash.ToArray().CopyTo(key, 1);
+
+            var contractStateBuffer = await stateAPI.GetStateAsync(
+                stateRoot.RootHash, NativeContract.ContractManagement.Hash, key);
+            var contractState = new StorageItem(contractStateBuffer).GetInteroperable<ContractState>();
+            var states = await rpcClient.ExpressFindStatesAsync(stateRoot.RootHash, _contractHash, new byte[0]);
+
+            return (contractState, states);
+        }
+
+        public static int PersistContract(SnapshotCache snapshot, ContractState state,
+            IReadOnlyList<(string key, string value)> storagePairs, ContractCommand.OverwriteForce force)
+        {
+            StorageKey key = new KeyBuilder(NativeContract.ContractManagement.Id, Prefix_Contract).Add(state.Hash);
+            var localContract = snapshot.GetAndChange(key)?.GetInteroperable<ContractState>();
+
+            if (localContract is null)
+            {
+                // Our local chain might already be using the contract id of the pulled contract, we need to check for this
+                // to avoid having contracts with duplicate id's. This is important because the contract id is part of the
+                // StorageContext used with Storage syscalls and else we'll potentially override storage keys or iterate
+                // over keys that shouldn't exist for one of the contracts.
+                if (state.Id <= LastUsedContractId(snapshot))
+                {
+                    state.Id = GetNextAvailableId(snapshot);
+                }
+                else
+                {
+                    // Update available id such that a regular contract deploy will use the right next id;
+                    SetNextAvailableContractId(snapshot, state.Id + 1);
+                }
+
+                snapshot.Add(key, new StorageItem(state));
+            }
+            else
+            {
+                if (force == ContractCommand.OverwriteForce.None)
+                {
+                    IList<(string key, string value)> states = new Collection<(string key, string value)>();
+                    byte[] prefixKey = StorageKey.CreateSearchPrefix(localContract.Id, new byte[]{});
+                    foreach (var (k, v) in snapshot.Find(prefixKey))
+                    {
+                        states.Add((Convert.ToBase64String(k.Key.ToArray()), Convert.ToBase64String(v.ToArray())));
+                    }
+                    
+                    var stateEquals = localContract.ToJson().ToByteArray(false)
```

Comment:
There is probably a better way of doing this, but at least it works (where `Equals` and `==` don't).